### PR TITLE
Provide integration with sneakers gem

### PR DESCRIPTION
**Pull request by aiomaster**
Hi,
I want to add integration with the sneakers gem.
Sneakers uses bunny to connect to a rabbitmq server and provides workers.
Testsetup would be difficult, so I didn't made any tests.

### CODE REVIEW

**Comment by kyrylo** [comment]
Thanks! Unit testing should be easy, see the shoryuken integration tests as a reference: https://github.com/airbrake/airbrake/blob/master/spec/unit/shoryuken_spec.rb

**Comment by kyrylo** [comment]
Alternatively, Sidekiq tests https://github.com/airbrake/airbrake/blob/master/spec/unit/sidekiq_spec.rb

**Comment by aiomaster** [comment]
It's not that easy I guess, but I'll try a little.

**Comment by kyrylo** [comment]
Thank you. Please also rebase on master (I just fixed the build).

**Comment by aiomaster** [comment]
While testing I've found, that the current sneakers version on github and the one on rubygems differ a little on the provided arguments to the error reporters.
I've updated the code, that it could work with both versions.

File: lib/airbrake/sneakers.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module Airbrake
+  module Sneakers
+    ##
+    # Provides integration with Sneakers.
+    #
+    # @see https://github.com/jondot/sneakers
+    class ErrorReporter
+      def call(exception, worker = nil, **context)
```
Comment (kyrylo) [inline lib/airbrake/sneakers.rb]: Is double asterisk intentional?

File: airbrake.gemspec [hunk 2]
```diff
@@ -40,6 +40,8 @@ DESC
   s.add_development_dependency 'appraisal', '~> 2'
   s.add_development_dependency 'rack', '~> 1'
   s.add_development_dependency 'webmock', '~> 2'
+  s.add_development_dependency 'sneakers', '~> 2'
+  s.add_development_dependency 'rr'
```
Comment (kyrylo) [inline airbrake.gemspec]: Why can't we use RSpec mocks (we use them for the rest of this project)?

Comment (aiomaster) [reply airbrake.gemspec]: Yes, we can. :)
I just copied some test setup from the sneakers gem and was lazy, but now I have rewritten it to use RSpec mocks like you suggested.

File: lib/airbrake/sneakers.rb (hunk 1 above)
Comment (kyrylo) [reply lib/airbrake/sneakers.rb]: Isn't it for keyword arguments? Why can't we use a single asterisk?

File: spec/unit/sneakers_spec.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+RSpec.describe Airbrake::Sneakers::ErrorReporter do
+  let(:worker) do
+    queue = instance_double('queue')
+    allow(queue).to receive_messages(
+      name: 'test-queue',
+      opts: {},
+      exchange: instance_double('exchange')
+    )
+
+    Sneakers.clear!
```
Comment (kyrylo) [inline spec/unit/sneakers_spec.rb]: We should probably call this in an `after` block. Last executed test will leave Sneakers "dirty".

File: spec/unit/sneakers_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    Sneakers.configure(daemonize: true, log: '/dev/null')
+    Sneakers::Worker.configure_metrics
```
Comment (kyrylo) [inline spec/unit/sneakers_spec.rb]: L15..L16 probably belong to a `before` block.

File: spec/unit/sneakers_spec.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    Class.new do
+      include Sneakers::Worker
+      from_queue 'defaults'
+
+      def work(_)
+        raise 'oops error'
+      end
+    end.new(queue, Concurrent::ImmediateExecutor.new)
```
Comment (kyrylo) [inline spec/unit/sneakers_spec.rb]: It seems like second parameter is optional. Can we omit it?

File: lib/airbrake/sneakers.rb (hunk 1 above)
Comment (aiomaster) [reply lib/airbrake/sneakers.rb]: I simply want to support the call without a given worker as second argument:
```ruby
opts = { foo: 'bar' }
ErrorReporter.call(exception, opts)
```
so now with double asterisk `worker` is `nil` and `context` is `{ foo: 'bar' }`.
If I had just used a single asterisk on `context` the options would get assigned to `worker` and `context` is `nil`.

They added the `worker` parameter and I want to provide a solution that works with and without a given `worker` as second parameter.

This only works if the given options hash has symbolized keys, but this is the case:
https://github.com/jondot/sneakers/blob/7c7697757af239e2232e34c31bfd48e8bdacd596/lib/sneakers/worker.rb#L72

Another way would be to check the type of the worker argument, but I think it is more ugly.
```ruby
def call(exception, worker, context = nil)
  if worker.is_a?(Hash)
    context = worker
    worker = nil
  end
  # ...
end
```

File: spec/unit/sneakers_spec.rb (hunk 5 above)
Comment (aiomaster) [reply spec/unit/sneakers_spec.rb]: it seems we can

File: spec/unit/sneakers_spec.rb (hunk 4 above)
Comment (aiomaster) [reply spec/unit/sneakers_spec.rb]: valid point

File: spec/unit/sneakers_spec.rb (hunk 3 above)
Comment (aiomaster) [reply spec/unit/sneakers_spec.rb]: true

File: lib/airbrake/sneakers.rb (hunk 1 above)
Comment (kyrylo) [reply lib/airbrake/sneakers.rb]: That makes sense. Can you please provide an example (just a link to the code) of a real job with keyword arguments?

Comment (kyrylo) [reply lib/airbrake/sneakers.rb]: By the way, I am fairly confident these details must be documented as a test. It's very tricky to understand so we must be sure that everything works as expected. Could you please a test(s) to make sure we support multiple arguments (different combinations).

Comment (aiomaster) [reply lib/airbrake/sneakers.rb]: The question is: Can or should we test multiple sneakers versions? It is easy to use the unreleased git version locally by adding `gem 'sneakers', github: 'jondot/sneakers'` to the Gemfile, but how could we manage that in a test?

I think I don't know, what you mean by "real job with keyword arguments". The keyword arguments are constructed here:
https://github.com/jondot/sneakers/blob/7c7697757af239e2232e34c31bfd48e8bdacd596/lib/sneakers/worker.rb#L72
and here the error_reporter is just called with it here:
https://github.com/jondot/sneakers/blob/7c7697757af239e2232e34c31bfd48e8bdacd596/lib/sneakers/error_reporter.rb#L18

But yeah, I can add a test, where the `Airbrake::Sneakers::ErrorReporter` is called directly with different arguments. Would that help?

**Comment by aiomaster** [comment]
Ok, I've added two more tests to call the ErrorReporter directly with different arguments.
Thanks for your reviews and your feedback!

File: lib/airbrake/sneakers.rb (hunk 1 above)
Comment (kyrylo) [reply lib/airbrake/sneakers.rb]: Testing this project is already hard enough. For simplicity we test only latest stable versions of libraries. So to answer your question, no we shouldn't.

Ok, I figured it out, let's stick to `**`.

**Comment by kyrylo** [comment]
👍 👍 thanks for the new integration!

**Comment by kyrylo** [comment]
Just tested on a Rails app, working good 👍

**Pull request merged**